src/server/auth.ts

```diff
@@ -53,3 +54,43 @@ export default async function initializePatronTokenAuth(req: NextRequest) {
 
   return patronTokenResponse
 }
+
+/**
+ * Creates and returns redirect url from initializePatronTokenAuth() response. Call this function only with an un-authenticated patron,
+ * i.e., patronTokenResponse.isTokenValid must be false.
+ */
+export function getLoginRedirect(req) {
+  const protocol = "http"
+  const hostname = req.headers["host"]
+  const originalUrl = BASE_URL + req.url
+  const fullUrl = encodeURIComponent(`${protocol}://${hostname}${originalUrl}`)
+  const redirect = `${
+    appConfig.externalUrls.loginUrl[appConfig.environment]
+  }?redirect_uri=${fullUrl}`
+  return redirect
+}
+
+/**
+ * Creates redirect to log out user, then return user to their current page.
+ */
+export const useLogoutRedirect = () => {
+  const baseUrl = "/research/research-catalog"
+  // Will send user back to prod if below doesn't work
+  const [redirect, setRedirect] = useState(`https://www.nypl.org${baseUrl}`)
+  useEffect(() => {
+    const current = window.location.pathname
+    let backPath = window.location.href
+    // If the patron is on any hold or account page, then
+    // redirect them to the home page after logging out. Otherwise,
+    // send them back to the page they were on.
+    if (current.includes("hold") || current.includes("account")) {
+      backPath = window.location.origin + baseUrl
+    }
+    setRedirect(
+      `${
+        appConfig.externalUrls.logoutUrl[appConfig.environment]
+      }?redirect_uri=${backPath}`
+    )
+  }, [])
```

Comment:
This works on all the pages I tested, but I'm also a bit confused how this does get updated even though it gets mounted once.

Reply:
why are we using window here instead of the next router object?

Reply:
Window allowed more specificity and makes it so we don't have to reconstruct the url from host, protocol, and pathname (which are not consistently available across all environments) but this does work with a router locally (see this commit https://github.com/NYPL/research-catalog/commit/e202e3b483ca2f2b1c42b9875f9fefa546e8776f)

Reply:
I'm surprised to learn about the inconsistency across environments with router.pathname. We're using that elsewhere so it might be good to reconsider its use in those places.

Reply:
what environments are they not available in?

Reply:
If it's the development environment where protocol is missing, it shouldn't be a problem for useRouter for most of the app. I think we rarely use it to determine the protocol.